app/controllers/CancelDeclarationController.scala

```diff
@@ -0,0 +1,126 @@
+  errorHandler: ErrorHandler
+)(implicit val messagesApi: MessagesApi) extends FrontendController with I18nSupport {
+
+  val formId = "cancelDeclarationForm"
+  val CANCEL_FUNCTION_CODE = 13
+
+  implicit val formats = Json.format[CancelDeclarationForm]
+
+  val form = Form(mapping(
+    "wcoDataModelVersionCode" -> optional(text()),
+    "wcoTypeName" -> optional(text()),
+    "responsibleAgencyName" -> optional(text()),
+    "functionalReferenceID" -> optional(text()),
+    "id" -> nonEmptyText,
+     "submitter" -> Submitter.formMapping,
+     "additionalInformation" -> AdditionalInformation.formMapping,
+      "amendment" -> Amendment.formMapping
+    )(CancelDeclarationForm.apply)(CancelDeclarationForm.unapply))
+
+  def displayForm(): Action[AnyContent] = authenticate.async { implicit request =>
+    customsCacheService.fetchAndGetEntry[CancelDeclarationForm](appConfig.appName, formId).map{
+      case Some(data) => Ok(cancelDeclaration(appConfig, form.fill(data)))
+      case _ =>  Ok(cancelDeclaration(appConfig, form))
+    }
+  }
+
+  def onSubmit(): Action[AnyContent] = authenticate.async { implicit request =>
+  form.bindFromRequest().fold(
+    (formWithErrors: Form[CancelDeclarationForm]) =>
+      Future.successful(BadRequest(cancelDeclaration(appConfig, formWithErrors))),
+    form => {
+      request.session
+      customsCacheService.cache[CancelDeclarationForm](appConfig.appName,formId,form).flatMap{ _ =>
+        customsDeclarationsConnector.submitCancellation(createCancellationMetadata(form)).flatMap{
+          case CustomsDeclarationsResponse(ACCEPTED, Some(conversationId)) =>
+            val submission = new Submission(request.user.eori, conversationId)
+            customsDeclareExportsConnector.saveSubmissionResponse(submission).flatMap{ _ =>
+              Future.successful(Ok(confirmation_page(appConfig, conversationId)))
+            }.recover{
+              case error: Throwable =>
+                Logger.error(s"Error from Customs Declare Exports ${error.toString}")
+                BadRequest(
+                  errorHandler.standardErrorTemplate(
+                    pageTitle = messagesApi("global.error.title"),
+                    heading = messagesApi("global.error.heading"),
+                    message = messagesApi("global.error.message")
+                  )
+                )
+            }
+          case error =>
+            Logger.error(s"Error from Customs declarations api ${error.toString}")
+            Future.successful(
+              BadRequest(
+                errorHandler.standardErrorTemplate(
+                  pageTitle = messagesApi("global.error.title"),
+                  heading = messagesApi("global.error.heading"),
+                  message = messagesApi("global.error.message")
+                )
+              )
+            )
+        }
+      }
+    }
+  )
+}
+
+  private def createCancellationMetadata(form:CancelDeclarationForm):MetaData =
```

Comment:
We can stay with this here, but I think that we should create services to put methods inside services, not in Controller. We have a lot of code in controllers that we shouldn't have.

Reply:
this method will not exist after we have UI navigation built